plugins/producer_plugin/include/eosio/producer_plugin/production_pause_vote_tracker.hpp

```diff
@@ -0,0 +1,260 @@
+   void force_unpause() {
+      latest_other_vote.reset_received_blocks();
+      latest_producer_vote.reset_received_blocks();
+   }
+
+   struct pause_status {
+      struct received_times {
+         fc::time_point                latest_vote{};
+         std::optional<fc::time_point> earliest_conflict_block{}; // If present, must be greater than latest_vote.
+      };
+
+      enum class pause_reason : unsigned int {
+         not_paused        = 0b00,
+         old_other_vote    = 0b01,
+         old_producer_vote = 0b10,
+         old_votes         = old_other_vote | old_producer_vote,
+      };
+
+      received_times other_vote{};
+      received_times producer_vote{};
+
+      pause_reason reason() const {
+         unsigned int r = static_cast<unsigned int>(pause_reason::not_paused);
+
+         if (other_vote.earliest_conflict_block.has_value()) {
+            r |= static_cast<unsigned int>(pause_reason::old_other_vote);
+         }
+
+         if (producer_vote.earliest_conflict_block.has_value()) {
+            r |= static_cast<unsigned int>(pause_reason::old_producer_vote);
+         }
+
+         return static_cast<pause_reason>(r);
+      }
+
+      bool should_pause() const {
+         return reason() != pause_reason::not_paused;
+      }
+   };
+
+   // Specify which vote timing check is needed in `check_pause_status`.
+   // If `producer`, then `other_vote` in returned `pause_status` is
+   // the default value and can be ignored.
+   // If `other`, then `producer_vote` in returned `pause_status` is
+   // the default value and can be ignored.
+   // Regardless, `should_pause()` and `reason()` called on returned
+   // `paused_status` is correct according to specified `pause_check`.
+   enum class pause_check { producer, other, both };
+
+   // Can be called concurrently with all member functions except:
+   //    + set_vote_timeout
+   //
+   // Returns a struct that provides useful time information tracked by this class
+   // which, most importantly, determines whether production should be paused.
+   // To determine whether production should be paused or not, simply check the boolean value
+   // returned from the should_pause member function of the returned struct.
+   pause_status check_pause_status(fc::time_point current_time, pause_check check = pause_check::both) const {
+      pause_status status;
+
+      const auto threshold_time = fc::time_point(current_time).safe_add(negative_vote_timeout);
+
+      auto process_vote_timing =
+         [&](const vote_timing& vt, pause_status::received_times& rt) {
+         const auto s = vt.get_vote_timing_status();
+
+         rt.latest_vote = s.latest_vote;
+
+         if (!s.first_block_after_vote.has_value())
+            return;
+
+         if (*s.first_block_after_vote < threshold_time) {
+            rt.earliest_conflict_block = *s.first_block_after_vote;
+         }
+      };
+
+      if (check != pause_check::producer) {
```

Comment:
Rather than using negative (`!=`) comparison, you can do it explicitly
 ```
if (both) {
   process vote for other;
   process vote timing for producer;
} else if (producer) {
   process vote for producer;
} else if (other) {
   process vote for other;
} else {
   assert;
}
```

This way you don't ned complicated asserts and code seems more clearer.

Reply:
Changed to switch.